File: Data Structures & Algorithms/reverse-a-linked-list/submission-7.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun reverseList(head: ListNode?): ListNode? {
        var previous: ListNode? = null
        var current = head

        while (current != null) {
            val temp = current.next
            current.next = previous
            previous = current
            current = temp
        }

        return previous
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-7.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        before = dummy
        after = dummy

        for _ in range(n):
            after = after.next

        while after.next:
            before = before.next
            after = after.next

        before.next = before.next.next

        return dummy.next